src/drivers/sifive_l2pf0.c

```diff
@@ -32,63 +30,45 @@
 #define REG_BITSHIFT_21 21
 #define REG_BITSHIFT_28 28
 
-/* Macros to capture trap, if L2PF does not exist for a hart id. */
-#define SIFIVE_L2PF0_TRAP_CAPTURE(exit, mtvec)                                 \
-    __asm__ __volatile__("la %0, 1f \n\t"                                      \
-                         "csrr %1, mtvec \n\t"                                 \
-                         "csrw mtvec, %0 \n\t"                                 \
-                         : "+r"(exit), "+r"(mtvec))
-
-#define SIFIVE_L2PF0_TRAP_RESTORE(mtvec)                                       \
-    __asm__ __volatile__(".align 2 \n\t"                                       \
-                         "1: \n\t"                                             \
-                         "csrw mtvec, %0 \n\t"                                 \
-                         : "+r"(mtvec))
+/* Array of base addresses with HART IDs as the index */
+unsigned long l2pf_base[] = METAL_SIFIVE_L2PF0_BASE_ADDR;
+int l2pf_base_len = sizeof(l2pf_base) / sizeof(unsigned long);
 
 void sifive_l2pf0_enable(void) {
-    volatile uintptr_t exit = 0, mtvec = 0;
     int hartid;
     __asm__ volatile("csrr %0, mhartid" : "=r"(hartid));
 
-    SIFIVE_L2PF0_TRAP_CAPTURE(exit, mtvec);
-
-    uint32_t val = REGW(METAL_SIFIVE_L2PF0_BASIC_CONTROL);
-
-    /* Enable L2 prefetch unit for current hart */
-    val |= REG_MASK_BITWIDTH1;
+    if (hartid < l2pf_base_len) {
```

Comment:
You are assuming the hartID are continuous and in sequence, which it may not.
But then the generator is assuming this too. Ideal a comment should be added onthis assumption, starting with "Alert!"